app/models/infra_conversion_job.rb

```diff
@@ -22,29 +22,50 @@ def self.create_job(options)
   #                                                    :finish             v
   #                                                                    finished
   #
+  # TODO: Update this diagram after we've settled on the updated state transitions.
 
   alias_method :initializing, :dispatch_start
   alias_method :finish,       :process_finished
   alias_method :abort_job,    :process_abort
   alias_method :cancel,       :process_cancel
   alias_method :error,        :process_error
 
+  # This is essentially an interface method called by the
+  # Job::StateMachine#transitions method.
+  #
   def load_transitions
     self.state ||= 'initialize'
 
     {
-      :initializing     => {'initialize'       => 'waiting_to_start'},
-      :start            => {'waiting_to_start' => 'running'},
-      :poll_conversion  => {'running'          => 'running'},
-      :start_post_stage => {'running'          => 'post_conversion'},
-      :poll_post_stage  => {'post_conversion'  => 'post_conversion'},
-      :finish           => {'*'                => 'finished'},
-      :abort_job        => {'*'                => 'aborting'},
-      :cancel           => {'*'                => 'canceling'},
-      :error            => {'*'                => '*'}
+      :initializing       => {'initialize'           => 'waiting_to_start'},
+      :start              => {'waiting_to_start'     => 'ready'},
+      :collapse_snapshots => {'ready'                => 'collapsing_snapshots', 'collapsing_snapshots' => 'collapsing_snapshots'},
+      :poll_conversion    => {'collapsing_snapshots' => 'running', 'running' => 'running' },
+      :start_post_stage   => {'running'              => 'post_conversion'},
+      :poll_post_stage    => {'post_conversion'      => 'post_conversion'},
+      :finish             => {'*'                    => 'finished'},
+      :abort_job          => {'*'                    => 'aborting'},
+      :cancel             => {'*'                    => 'canceling'},
+      :error              => {'*'                    => '*'}
     }
   end
 
+  # Collapse (delete) the snapshots of the virtual machine, because disk-sync
+  # doesn't work if the VM has snapshots. This is a limitation of CBT.
+  #
+  def collapse_snapshots
+    message = 'Collapse Snapshots'
+    update(:message => message)
+
+    if migration_task.source.supports_remove_all_snapshots?
+      _log.info(prep_message(message))
+      migration_task.source.remove_all_snapshots_queue(migration_task.userid)
+    end
+
+    handover_to_automate
+    queue_signal(:poll_conversion)
```

Comment:
I would expect it to `queue_signal(:collapse_snapshots)` if the task isn't finished. You probably need to store the task id in the job context.

Reply:
I guess I need to understand why this shouldn't be a blocking method, since we can't proceed unless the snapshots have been deleted, can we?

Reply:
I personally prefer non blocking methods, because you release some time for other jobs. But I confess that it's an habit from Automate development, rather than factual :)

Reply:
@fdupont-redhat Ok, took an initial stab at it. Not sure if that's the right approach or not.

Reply:
Doesn't it start a new task on each iteration? Maybe you should store the task id in `context` and check for its existence. Also, if the task fails, we should abort because snapshots removal is a prerequisite.

Reply:
I don't think I'm understanding how you want this written, so I'll let you update it when you're back on Monday. :)

Reply:
A proposal is here: https://github.com/fdupont-redhat/manageiq/blob/v2v_state_machine_collapse_snapshots/app/models/infra_conversion_job.rb.

It uses the fact that we have an async task to stay in the 'collapsing_snapshot' state and retry later. We need to store the task id in the job context, so that when we retry we can retrieve the task status and decide what to do: 1. retry if it's not finished, 2. abort if it failed, 3. call next state if it succeeded.

Reply:
@fdupont-redhat Looks like you deleted it. :(